airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/job_tracker/JobTracker.java

```diff
@@ -168,30 +179,130 @@ public void trackSync(final Job job, final JobState jobState) {
+  protected static Map<String, Object> configToMetadata(final String jsonPath, final JsonNode config, final JsonNode schema) {
+    final Map<String, Object> metadata = configToMetadata(config, schema);
+    // Prepend all the keys with the root jsonPath
+    // But leave the values unchanged
     final Map<String, Object> output = new HashMap<>();
+    mergeMaps(output, jsonPath, metadata);
+    return output;
+  }
 
-    if (config.isObject()) {
-      final ObjectNode node = (ObjectNode) config;
-      for (final Iterator<Map.Entry<String, JsonNode>> it = node.fields(); it.hasNext();) {
-        final var entry = it.next();
-        final var field = entry.getKey();
-        final var fieldJsonPath = jsonPath + "." + field;
-        final var child = entry.getValue();
-
-        if (child.isBoolean()) {
-          output.put(fieldJsonPath, child.asBoolean());
-        } else if (!child.isNull()) {
-          if (child.isObject()) {
-            output.putAll(configToMetadata(fieldJsonPath, child));
-          } else if (!child.isTextual() || (child.isTextual() && !child.asText().isEmpty())) {
-            output.put(fieldJsonPath, SET);
+  /**
+   * Does the actually interesting bits of configToMetadata. If config is an object, returns a
+   * flattened map. If config is _not_ an object (i.e. it's a primitive string/number/etc, or it's an
+   * array) then returns a map of {null: toMetadataValue(config)}.
+   */
+  private static Map<String, Object> configToMetadata(final JsonNode config, final JsonNode schema) {
+    if (schema.hasNonNull("const")) {
+      // If this schema is a const, then just dump it into a map:
+      // * If it's an object, flatten it
+      // * Otherwise, do some basic conversions to value-ish data.
+      // It would be a weird thing to declare const: null, but in that case we don't want to report null
+      // anyway, so explicitly use hasNonNull.
+      return flatten(config);
+    } else if (schema.has("oneOf")) {
+      // If this schema is a oneOf, then find the first sub-schema which the config matches
+      // and use that sub-schema to convert the config to a map
+      final JsonSchemaValidator validator = new JsonSchemaValidator();
+      for (final Iterator<JsonNode> it = schema.get("oneOf").elements(); it.hasNext();) {
+        final JsonNode subSchema = it.next();
+        if (validator.test(subSchema, config)) {
+          return configToMetadata(config, subSchema);
+        }
+      }
+      // If we didn't match any of the subschemas, then something is wrong. Bail out silently.
+      return emptyMap();
+    } else if (config.isObject()) {
+      // If the schema is not a oneOf, but the config is an object (i.e. the schema has "type": "object")
+      // then we need to recursively convert each field of the object to a map.
+      // Note: this doesn't handle additionalProperties, so if a config contains properties that are not
+      // explicitly declared in the schema, they will be ignored.
```

Comment:
Would it be worth reporting the names of `additionalProperties` instead of ignoring them?

Reply:
realized that I might as well just handle additionalProperties in a smarter way, can you take a look at this? https://github.com/airbytehq/airbyte/pull/10120/commits/e74066d149529dae4575d81688f49460637d5708#diff-8ef120819d2dd157a1e8d1838a49917772872273fefb9719b56aa303174581abL235-R261

Reply:
LGTM!